### PR TITLE
Routine updates

### PR DESCRIPTION
**Pull request by jglick**
Bring this into the current world.

@reviewbybees

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -3,11 +3,12 @@
...
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>2.24</version>
+    <version>3.2</version>
+    <relativePath />
   </parent>
 
   <artifactId>distfork</artifactId>
-  <version>1.6.1-SNAPSHOT</version>
+  <version>1.7-SNAPSHOT</version>
```
Comment (jglick) [inline pom.xml:11]: no need for micro releases I think

File: pom.xml [hunk 2]
```diff
@@ -17,10 +18,11 @@
                instead of hardcoding a machine name, you can specify a label to let Jenkins choose a agent.
                This opens up a Jenkins cluster for doing all kinds of things without requiring a job/build notion.
   </description>
-  <url>http://wiki.jenkins-ci.org/display/JENKINS/DistFork+Plugin</url>
+  <url>https://wiki.jenkins.io/display/JENKINS/DistFork+Plugin</url>
 
   <properties>
-    <findbugs.failOnError>false</findbugs.failOnError>
+    <jenkins.version>2.60.3</jenkins.version>
```
Comment (jglick) [inline pom.xml:24]: Just makes things easier.

File: pom.xml [hunk 3]
```diff
@@ -30,42 +32,31 @@
     <tag>HEAD</tag>
   </scm>
 
-  <developers>
-    <developer>
-      <id>kohsuke</id>
-      <name>Kohsuke Kawaguchi</name>
```
Comment (jglick) [inline pom.xml:36]: does not seem to be true any more

File: src/main/java/hudson/plugins/distfork/DistForkCommand.java [hunk 4]
```diff
@@ -125,7 +119,8 @@ protected int run() throws Exception {
             // l.getNodes() includes the master if this has the specified label
             // but j.getNodes() does not include the master so check explicitly.
             if (!hasPermission && j.getNumExecutors() > 0) {
-                hasPermission = j.toComputer().hasPermission(Computer.BUILD); 
+                Computer c = j.toComputer();
+                hasPermission = c != null && c.hasPermission(Computer.BUILD);
```
Comment (jglick) [inline src/main/java/hudson/plugins/distfork/DistForkCommand.java:123]: FindBugs

File: src/main/java/hudson/plugins/distfork/DistForkCommand.java [hunk 5]
```diff
@@ -159,9 +154,17 @@ public void run() {
                 try {
                     Computer c = Computer.currentComputer();
                     Node n = c.getNode();
-                    String nodeName = n.getNodeName().equals("") ? "master" : n.getNodeName();
+                    if (n == null) {
+                        throw new IllegalStateException("missing node " + c.getName());
```
Comment (jglick) [inline src/main/java/hudson/plugins/distfork/DistForkCommand.java:158]: ditto

File: src/main/java/hudson/plugins/distfork/DistForkCommand.java [hunk 6]
```diff
@@ -159,9 +154,17 @@ public void run() {
...
+                    }
+                    String nodeName = n.getNodeName().isEmpty() ? "master" : n.getNodeName();
```
Comment (jglick) [inline src/main/java/hudson/plugins/distfork/DistForkCommand.java:160]: IDE hint

File: src/main/java/hudson/plugins/distfork/DistForkCommand.java [hunk 7]
```diff
@@ -159,9 +154,17 @@ public void run() {
...
                     listener.getLogger().println("Executing on " + nodeName);
-                    FilePath workDir = n.getRootPath().createTempDir("distfork",null);
+                    FilePath root = n.getRootPath();
+                    if (root == null) {
+                        throw new IllegalStateException(nodeName + " seems to be offline");
```
Comment (jglick) [inline src/main/java/hudson/plugins/distfork/DistForkCommand.java:164]: FindBugs

File: src/main/java/hudson/plugins/distfork/DistForkCommand.java [hunk 8]
```diff
@@ -159,9 +154,17 @@ public void run() {
...
+                    }
+                    root.mkdirs(); // needed at least for mock-slave
```
Comment (jglick) [inline src/main/java/hudson/plugins/distfork/DistForkCommand.java:166]: after `mock-slave` update

File: src/main/java/hudson/plugins/distfork/DistForkCommand.java [hunk 9]
```diff
@@ -238,8 +241,11 @@ private void setUpPortForwarding(List<PortSpec> fowrarding, VirtualChannel recv,
         });
 
         // run and wait for the completion
-        Queue q = j.getQueue();
-        Future<Executable> f = q.schedule(t, 0).getFuture();
+        Queue.WaitingItem item = j.getQueue().schedule(t, 0);
+        if (item == null) {
+            throw new IllegalStateException("Could not schedule task");
```
Comment (jglick) [inline src/main/java/hudson/plugins/distfork/DistForkCommand.java:246]: FindBugs

File: src/test/java/hudson/plugins/distfork/DistForkCommandTest.java [hunk 10]
```diff
@@ -91,7 +107,7 @@ public void testAnonymousAccess() throws Exception {
         jr.jenkins.setAuthorizationStrategy(authz);
         
         String result = commandAndOutput("dist-fork", "-l", "master", "whoami");
-        assertThat(result, containsString("hudson.security.AccessDeniedException2: anonymous is missing the Slave/Build permission"));
+        assertThat(result, containsString(new AccessDeniedException2(Jenkins.ANONYMOUS, Computer.BUILD).getMessage()));
```
Comment (jglick) [inline src/test/java/hudson/plugins/distfork/DistForkCommandTest.java:115]: now `Agent/Build`, and exception class name no longer printed (IIRC part of the general cleanup of CLI return codes)

File: src/test/java/hudson/plugins/distfork/DistForkCommandTest.java [hunk 11]
```diff
@@ -133,12 +149,14 @@ public void testUserWithOutBuildAccess() throws Exception {
...
-        assertThat(result, containsString("hudson.security.AccessDeniedException2: alice is missing the Slave/Build permission"));
+        assertThat(result, containsString(new AccessDeniedException2(User.getById("alice", false).impersonate(), Computer.BUILD).getMessage()));
     }
-    
+
+    @Ignore("TODO against baseline when written, this was meaningless (bob was an admin); as of JENKINS-37616, it fails in the expected way: the task hangs in the queue because you need Computer.BUILD")
     @Test
     @Issue("SECURITY-386")
     public void testUserWithProvisionAccess() throws Exception {
+        logging.record(Queue.class, Level.FINEST);
```
Comment (jglick) [inline src/test/java/hudson/plugins/distfork/DistForkCommandTest.java:164]: very useful for tracking down provisioning issues

File: src/test/java/hudson/plugins/distfork/DistForkCommandTest.java [hunk 12]
```diff
@@ -155,7 +173,7 @@ public void testUserWithProvisionAccess() throws Exception {
         jr.jenkins.setSecurityRealm(realm);
         jr.jenkins.setAuthorizationStrategy(authz);
         
-        jr.jenkins.clouds.add(new MockCloud("Mock Cloud", Mode.NORMAL, 1, "cloud"));
+        jr.jenkins.clouds.add(new MockCloud("Mock Cloud", Mode.NORMAL, 1, "cloud", true));
```
Comment (jglick) [inline src/test/java/hudson/plugins/distfork/DistForkCommandTest.java:181]: the default for `executors == 1`; does not seem to matter either way

File: src/test/java/hudson/plugins/distfork/DistForkCommandTest.java [hunk 13]
```diff
@@ -164,7 +182,8 @@ public void testUserWithProvisionAccess() throws Exception {
     @Test
     @Issue("SECURITY-386")
     public void testUserWithBuildAccessOnCloud() throws Exception {
-        // a user without Computer.BUILD should be able to run this command.
+        logging.record(Queue.class, Level.FINEST);
+        // a user without Cloud.PROVISION should be able to run this command.
```
Comment (jglick) [inline src/test/java/hudson/plugins/distfork/DistForkCommandTest.java:191]: I think mistyped in the original

File: src/test/java/hudson/plugins/distfork/DistForkCommandTest.java [hunk 14]
```diff
@@ -133,12 +149,14 @@ public void testUserWithOutBuildAccess() throws Exception {
         jr.jenkins.setAuthorizationStrategy(authz);
         
         String result = commandAndOutput("dist-fork", "--username=alice", "--password=alice", "-l", "master", "whoami");
-        assertThat(result, containsString("hudson.security.AccessDeniedException2: alice is missing the Slave/Build permission"));
+        assertThat(result, containsString(new AccessDeniedException2(User.getById("alice", false).impersonate(), Computer.BUILD).getMessage()));
     }
-    
+
+    @Ignore("TODO against baseline when written, this was meaningless (bob was an admin); as of JENKINS-37616, it fails in the expected way: the task hangs in the queue because you need Computer.BUILD")
```
Comment (jglick) [inline src/test/java/hudson/plugins/distfork/DistForkCommandTest.java:160]: cf. https://github.com/jenkinsci/jenkins/pull/2521

File: src/test/java/hudson/plugins/distfork/DistForkCommandTest.java [hunk 15]
```diff
@@ -37,21 +37,37 @@
...
     @Rule
     public JenkinsRule jr = new JenkinsRule();
 
+    @Rule
+    public LoggerRule logging = new LoggerRule();
+
+    /** JENKINS_24752: otherwise {@link #testUserWithBuildAccessOnCloud} waits a long time */
+    @BeforeClass
+    public static void runFaster() {
+        System.setProperty("hudson.model.LoadStatistics.clock", "1000");
```
Comment (jglick) [inline src/test/java/hudson/plugins/distfork/DistForkCommandTest.java:69]: picked up from `workflow-aggregator-plugin/demo/` etc.

File: src/main/java/hudson/plugins/distfork/DistForkCommand.java [hunk 16]
```diff
@@ -269,6 +275,7 @@ private static final boolean hasBuildPermission(Collection<? extends Node> nodes
 
     /**
      * Check if the current user has permission to provision on any of the specified clouds.
+     * TODO as of 2.20 this is useless and the task will just hang anyway (see test)
```
Comment (jglick) [inline src/main/java/hudson/plugins/distfork/DistForkCommand.java:278]: could perhaps just be deleted, but erring on the side of caution here

File: pom.xml [hunk 17]
```diff
@@ -30,42 +32,31 @@
...
-    </developer>
-  </developers>
-
   <dependencies>
-    <dependency>
+    <dependency> <!-- TODO switch to MockAuthorizationStrategy -->
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>matrix-auth</artifactId>
-      <version>1.4</version>
+      <version>2.2</version>
```
Comment (jglick) [inline pom.xml:39]: forces 2.60.x but simplifies `requireUpperBoundsDep`

File: pom.xml [hunk 18]
```diff
@@ -17,10 +18,11 @@
                instead of hardcoding a machine name, you can specify a label to let Jenkins choose a agent.
                This opens up a Jenkins cluster for doing all kinds of things without requiring a job/build notion.
   </description>
-  <url>http://wiki.jenkins-ci.org/display/JENKINS/DistFork+Plugin</url>
+  <url>https://wiki.jenkins.io/display/JENKINS/DistFork+Plugin</url>
```
Comment (jglick) [inline pom.xml:21]: canonical URL

File: src/main/java/hudson/plugins/distfork/DistForkCommand.java (hunk 16 above)
Comment (oleg-nenashev) [reply src/main/java/hudson/plugins/distfork/DistForkCommand.java:278]: I would keep it as is for now

**Review by oleg-nenashev — APPROVED** [review]
🐝

**Pull request merged**